src/common/events/events_io.cpp

```diff
@@ -137,6 +137,76 @@ mio::ExpectedSize ReaderFromAsyncReader::Read(
 	return read;
 }
 
+TeeReader::TeeReaderLeafPtr TeeReader::AttachSecondaryReader() {
+	auto reader = make_shared<TeeReaderLeaf>(shared_from_this());
+	leaf_readers_.insert({reader, TeeReaderRequestedRead {}});
+	return reader;
+}
+
+error::Error TeeReader::ReadyToAsyncRead(
+	TeeReader::TeeReaderLeafPtr leaf_reader,
+	vector<uint8_t>::iterator start,
+	vector<uint8_t>::iterator end,
+	mio::AsyncIoHandler handler) {
+	// TODO: should this be an assert() instead?
+	auto found = leaf_readers_.find(leaf_reader);
+	if (found == leaf_readers_.end()) {
+		return error::MakeError(
+			error::ProgrammingError, "Leaf reader not found in internal map. This cannot happen");
+	}
+
+	leaf_readers_[leaf_reader].start = start;
+	leaf_readers_[leaf_reader].end = end;
+	leaf_readers_[leaf_reader].handler = handler;
+
+	if (++ready_to_read == leaf_readers_.size()) {
+		DoAsyncRead();
+		// TODO: should this be done from the _handler_ in DoAsyncRead()?
+		ready_to_read = 0;
+	}
+
+	return error::NoError;
+}
+
+void TeeReader::DoAsyncRead() {
+	auto handler = [this](mio::ExpectedSize result) {
+		if (!result) {
+			for (auto const &it : leaf_readers_) {
+				it.second.handler(result);
+			}
+			return;
+		};
+
+		for (auto &it : leaf_readers_) {
+			std::copy_n(buffer_.begin(), result.value(), it.second.start);
+		}
+		buffer_.clear();
+
+		for (auto const &it : leaf_readers_) {
+			it.second.handler(result);
+		}
+	};
+
+	auto min_read = std::min_element(
+		leaf_readers_.cbegin(),
+		leaf_readers_.cend(),
+		[](const std::pair<TeeReaderLeafPtr, TeeReaderRequestedRead> r1,
+		   std::pair<TeeReaderLeafPtr, TeeReaderRequestedRead> r2) {
+			return (r1.second.end - r1.second.start) < (r2.second.end - r2.second.start);
+		});
+	buffer_.resize(min_read->second.end - min_read->second.start);
```

Comment:
Why do we even need a buffer here? Can't you just read directly into the iterator range of the first reader, and then copy to the other afterwards?

Reply:
Thank you. I actually thought about this with some "not staring to pixels" time :)